### ISSUE TITLE
Has a big time gap between GPU and CPU for Rearrange op.

### ISSUE THREAD

**Issue by yl305237731** [arogozhnikov/einops#173]
for [vit.pytorch](https://github.com/lucidrains/vit-pytorch/blob/main/vit_pytorch/vit.py), the code of to_patch_embedding:
      `         self.to_patch_embedding = nn.Sequential(
            Rearrange('b c (h p1) (w p2) -> b (h w) (p1 p2 c)', p1 = patch_height, p2 = patch_width),
            nn.Linear(patch_dim, dim),
        )`

I test their inference time on cpu and gpu,

        'def forward(self, img):
        begin = time.time()
        x = self.to_patch_embedding(img)
        print('time of to_patch_embedding {}'.format(time.time() - begin))
        b, n, _ = x.shape
        cls_tokens = repeat(self.cls_token, '() n d -> b n d', b = b)
        x = torch.cat((cls_tokens, x), dim=1)
        begin_pos = time.time()
        x += self.pos_embedding[:, :(n + 1)]
        print('time of pos_embedding {}'.format(time.time() - begin_pos))
        x = self.dropout(x)
        begin_trans = time.time()
        x = self.transformer(x)
        print('time of transformer {}'.format(time.time() - begin_trans))'

for cpu infer,
![image](https://user-images.githubusercontent.com/9928596/155833251-9312f4ee-1b98-41b9-b3f1-b13030653c7d.png)

for gpu infer
![image](https://user-images.githubusercontent.com/9928596/155833291-1d218b4b-9359-463f-a221-9bd1bf88198d.png)

When using GPU, to_ path_ Embedding is much slower. How can we improve it?

**Comment by arogozhnikov** [comment]
Hi @yl305237731 , interesting question, difference in time you see it large (even, ridiculously large), but I don't think this is due to einops.

In particular: einops doesn't know about device being used, it's code and overhead does not depend on cpu/gpu, so overhead can't be greater than for CPU (thus, less than 0.0033 seconds which also include linear layer).

I'd recommend checking your GPU setup, it may need some time to start, or to perform memory management, or to transfer/compute image tensor or transfer model weights. To explain the last point: when you send a tensor to device, the code does not wait, you need to sync cuda in multiple places to have clear picture of what really takes time). 
Other helpful tests are running multiple times and converting model to torch.jit.traced for an inference.

**Comment by yl305237731** [comment]
@arogozhnikov thanks, you are right, this big time gap caused by nn.linear, although I don't know why at present.
